### ISSUE TITLE
Add warning before posting a media without alt-text

### ISSUE THREAD

**Issue by ffffluffy** [mastodon/mastodon#5644]
With #117, Mastodon now allows users to add an alternative text to media in order to improve accessibility for visually impaired people. However, many people, including me, too often forget to do so. I think it would be great to display a warning when a media doesn't have alt-text yet, before the user send the toot.

I was thinking of a warning such as "The media doesn't have alt-text", displayed either on the media (in a similar way than alt-text), or a banner above or beyond the text area.

I don't mean to make a mandatory system, but a reminder would be really helpful.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by jennifermf** [comment]
Yes please! I love this idea. Not mandatory, but a reminder :)

**Comment by wxcafe** [comment]
👍

**Comment by wimvanderbauwhede** [comment]
There was a thread about this yesterday with quite a few people (7) in favour of this:
https://octodon.social/web/statuses/99662063368952277
So I'd like to add their voice to this issue.

**Comment by timmc** [comment]
I'd be in favor of mandatory alt text, tbh. That's a much stronger nudge to make sure content on Mastodon is accessible to everyone. If you *really* don't want to put *any* description, even "comic" or "photo" or something, you can still put some dots or whatever. But I think it would be great for getting people to actually add alt text, especially because it's so easy to forget.

(So many images are of text, and it's a shame that that text is then not copied into the alt text... or the toot, for that matter...)

**Comment by nightpool** [comment]
i'm :-1: on mandatory validation, but :+1: on a warning indicator.

requiring EVERYBODY who posts an image to have the verbal spoons to caption that image is just trading one type of inaccessibility for another.

**Comment by timmc** [comment]
I'm in fact saying they don't have to have those spoons. They can use "-" if it's just too much. It's still in some ways just a warning, but it better sets a social standard.

**Comment by nightpool** [comment]
i feel like most well intentioned people would just give up posting rather then add a misleading alt text, and that sounds bad.

**Comment by jennifermf** [comment]
I choose not to post photos if I don't have the energy or mental fortitude to provide at least a rudimentary alt-text (like "sleeping cat"). I don't think this is bad, i think this is what building an inclusive community looks like. I think it's bad to knowingly and willfully exclude people from from something because jotting down a couple words is too cumbersome.

as @timmc says, folks can just enter a single character rather than a full description-- to at least acknowledge that they are excluding people from a post.

i previously said above that I was not in favor of a mandatory reminder. But this thread is reminding me that mastodon already has an enormous accessibility problem that extends well beyond alt-text options, so now I am changing my opinion: yes, it should be mandatory.

the philosophy that accessibility should only exist if it doesn't inconvenience abled people is an able-bodied and able-minded tactic to silence and exclude the disabled. i hope it's obvious why this is the antithesis of building a strong and inclusive fediverse.

**Comment by timmc** [comment]
I'm sure we can figure out some way to not be hostile to *either* people with low-vision/shitty connection/etc. or newcomers who are confused about their options for alt-text. Required, but if they try to skip it? hint that "-" or similar will let it by? I'm not very good with UI design, unfortunately, so I'm hoping someone else will be more clever. :-P

**Comment by nightpool** [comment]
@jennifermf the idea that two people's disabilities can cause conflicting access needs is not "the philosophy that accessibility should only exist if it doesn't inconvenience abled people". it's the exact opposite of that, and it's activism 101 stuff.

i support a reminder. i don't support a mandatory requirement. mastodon should be for everyone, not just those who have the energy to present themselves in a legible way 100% of the time.

**Comment by Gargron** [comment]
Mastodon is built on the social contract between people following each other: If you expect your followers to expect you to write alt text, you do that; If you don't, you don't. I'm not in favour of "hardcoding" such social relations. However you can enforce this in your code of conduct if you want.

**Comment by Cassolotl** [comment]
I once read a twitter thread about accessibility and image captions for screen readers. The person basically said that if the content of the picture is clear from the text in the body of the tweet (eg: "check out this cute doggo I saw at the train station today!") then an image description is just not required, and might actually be a waste of time for the visually impaired person.

I don't think saying "people can just put a hyphen" is a good solution; it feels messy. I think we have to trust people to judge when an image description is and isn't relevant or helpful.

**Comment by Sciss** [comment]
@Gargron as it stands in the current version I'm seeing (2.3.3), it's totally not obvious to add the caption. I know people who do have problems when there are no captions; so of course, one should be free not to add a caption, but the UI should definitely be more helpful and obviously indicating that there is no caption. I personally would caption all my image uploads, but I simply _forget_ all the time, because the UI does not indicate anything.

**Comment by penduin** [comment]
at least on mobile, the description field is hidden by default, but should absolutely be shown, i'd say even focused upon inserting an image.

there are several user-facing options to confirm when boosting etc, i'd love to see one to confirm posting with alt-less embedded media.

**Comment by ZoeBijl** [comment]
> Mastodon is built on the social contract between people following each other: If you expect your followers to expect you to write alt text, you do that; If you don't, you don't.

Sometimes one simply forgets to add alt text. Of course “Delete & Re-draft” helps a lot here. Would still be better if you got a warning about alt text tho. Of course community maintainers also need to inform their peers about the importance of alt text. A warning in Mastodon itself might help reach more people.

> I'm not in favour of "hardcoding" such social relations. However you can enforce this in your code of conduct if you want.

Would it be possible to make this an option admins/users can disable in their Settings?

**Comment by Sciss** [comment]
Also, it must be clearly said that Masto 3 is a regression in this case - before you could simply click on the image and start typing, now I need to open a separate dialogue to enter the alt-text, making it even more unlikely that people will do this, because of the added effort.

**Comment by Cassolotl** [comment]
>Also, it must be clearly said that Masto 3 is a regression in this case - before you could simply click on the image and start typing, now I need to open a separate dialogue to enter the alt-text, making it even more unlikely that people will do this, because of the added effort.

I *emphatically* disagree.

Before Mastodon 3, there was a tiny text window in which to type your description. You could see maybe under 30 characters at once? (Screenshots: https://github.com/tootsuite/mastodon/issues/6868) And that reduces if you attach more images. You couldn't scroll it or anything, you couldn't see the image you were describing, it was a nightmare.

This is how I would add alt text before 3.0:

1. Upload photo to Mastodon, and put the cursor in the alt text box.
2. Find the file I uploaded on my computer, open the image in Preview.
3. Open a new TextEdit window to type in.
4. Resize windows so that I can see the Preview window and the TextEdit window at the same time.
5. Type the image description in TextEdit while looking at the image in Preview.
6. Copy and paste the image description from TextEdit to the alt text box on the image in Mastodon.

***... for each image I was attaching to the post.***

Now you can see the whole image and all the text at once. Since I don't have to use three programs at once to comfortably add an image description, I am more likely to enter alt text.

**Comment by Pixdigit** [comment]
Since some people prefer being reminded to add captions and some not, why not add a setting? I know adding settings makes the settings more cluttered and especially with such detailed settings can become quite confusing. But as evidenced by this and the various other threads across the Fediverse this is a feature that many users want. Another idea worth debating would be a server wide setting; allowing admins to configure and enforce how they expect their community to interact.

**Comment by mkljczk** [comment]
What about #15590? I made it configurable, disabled by default.

**Comment by rkingett** [comment]
I like this idea more than a hard requirement. If it really bothers people that much, then why not set it so that instance administrators could make this an optional setting for all users, otherwise, the reminder is on by default depending on what instance you stay in.

**Comment by Be-ing** [comment]
> Since some people prefer being reminded to add captions and some not, why not add a setting?

Because Mastodon already has a ton of settings and its configuration UI is rather overwhelming. I don't think there should be a setting for this. Just opt out when you intentionally post without a caption.

**Comment by devinprater** [comment]
I hope this becomes reality. Even a reminder to add alt-text in the image upload dialog, so one could first upload them image, then write a description of it when it's done uploading, and the image shows up above the text box hopefully. As a blind user, I love that Pinafore, a much more accessible Mastodon client, has this I think. And Tusky, an accessible Mastodon client for Android, even tells me if an image is described or not before I even focus on it, so I know to skip the Toot or not. I mean, if we want Mastodon to be better than Twitter, let's make it even more accessible than Twitter also, including giving people the reminder they want/need in order to generate a more accessible environment.

**Comment by Be-ing** [comment]
Because discussion on this topic has gotten a bit scattered, I'll copy and paste my comment from #15590 here:

> I think we can come up with a smoother way to handle this than adding a modal dialog to the existing design. How about adding the textbox for editing the description to the posting form below the image thumbnail? A short text string could be shown below the "TOOT" button while the description textbox is empty, for example "Consider adding image description". This would be a clear reminder without getting in the way of users who intentionally want to post an image without a description. There would be no need to add a setting for this.
> Thanks to @byronalley for this [idea](https://github.com/tootsuite/mastodon/issues/13894#issuecomment-640737181).

**Comment by Be-ing** [comment]
@Gargron:
> Mastodon is built on the social contract between people following each other: If you expect your followers to expect you to write alt text, you do that; If you don't, you don't. I'm not in favour of "hardcoding" such social relations. However you can enforce this in your code of conduct if you want.

Would you be okay with merging a PR implementing my proposal above that doesn't add any extra steps for users who choose to post images without descriptions, only add a nonobstructive reminder?

**Comment by rkingett** [comment]
Twitter just released a feature where they alert users, aggressively, if they forget to add alt text to images and many like it. They use a modal dialog the user has to actively dismiss. It was just released but is praised already.

This should be aggressive as well

**Comment by Be-ing** [comment]
> This should be aggressive as well

I disagree. I don't think software should be aggressive. I think a gentle reminder that doesn't get in the way of people who want to ignore it would be effective without needing to add [yet another option](https://www.youtube.com/watch?v=glZ1C-Yu5tw) to the preferences.

**Comment by rkingett** [comment]
Twitter is starting to roll this out, globally, so why can't we have something like this either on the users side or an instance side? Like an administrator could enable this reminder by default for their instance?

**Comment by Be-ing** [comment]
Please do not give admins the option to turn this off.

**Comment by rkingett** [comment]
I'd also like it to where if they enable the instance image reminders for their instance, they would be unable to turn it off. Twitter is currently testing this on their platform so why can't we do it here too?

**Comment by devinprater** [comment]
I mean, I own an instance that I pay for to be maintained. I'd love an "accessibility" section in admin preferences, where I can not only say that media should be described, but that I should *only* get described media in my timelines. We should have control, not the simplistic "non-forceful" view that enables ableism.

**Comment by mkljczk** [comment]
Fixed by #17845?

**Comment by Be-ing** [comment]
> Fixed by https://github.com/mastodon/mastodon/pull/17845?

Looks like it.

**Comment by cirosantilli** [comment]
This warning is extremely annoying, I wish I could get rid of it and be allowed to post images without alt text if I want to without being asked every time like every other website. I know it is not ideal for accessibility, but I'm not going to add alt text because the warning tells me too, I'm too busy already procrastinating my life in other ways.

**Comment by Gargron** [comment]
![Image](https://github.com/user-attachments/assets/9973a95c-8bbd-4b05-911c-9044b32e702a)
